src/components/TextEditor/components/TextEditor.js

```diff
@@ -1,164 +1,141 @@
-  constructor(props) {
-    super(props);
+function TextEditor(props) {
+  const [codeMirrorInstance, setCodeMirrorInstance] = useState(null);
+  const [currentLine, setCurrentLine] = useState(0);
+  // const [sketch, setSketch] = useState(null);
+  const [showForkModal, setShowForkModal] = useState(false);
+  const [forking, setForking] = useState(false);
+  const [forked, setForked] = useState(false);
+  const [redirectToSketch, setRedirectToSketch] = useState(false);
+  const [showShareModal, setShowShareModal] = useState(false);
+  // const [error, setError] = useState(null);
 
-    this.state = {
-      codeMirrorInstance: null,
-      currentLine: 0,
-      sketch: null,
-      showForkModal: false,
-      forking: false,
-      forked: false,
-      redirectToSketch: false,
-      showShareModal: false,
-    };
-  }
+  const {
+    dirty,
+    mostRecentProgram,
+    code,
+    dirtyCode,
+    setProgramCode,
+    vthumbnail,
+    vlanguage,
+    sketchName,
+    uid,
+    theme,
+    thumbnail,
+    addProgram,
+    viewMode,
+    viewOnly,
+    updateViewMode,
+    screenHeight,
+    screenWidth,
+    language,
+    handleSave,
+    handleDownload,
+    saveText,
+  } = props;
 
   //= =============React Lifecycle Functions===================//
 
-  componentDidMount() {
-    window.addEventListener('beforeunload', this.onLeave);
-    window.addEventListener('close', this.onLeave);
-  }
-
-  componentWillUnmount = () => {
-    window.removeEventListener('beforeunload', this.onLeave);
-    window.removeEventListener('close', this.onLeave);
-  };
-
-  openForkModal = () => {
-    this.setState({ showForkModal: true });
-  };
-
-  closeForkModal = () => {
-    this.setState({ showForkModal: false });
+  const onLeave = async (ev) => {
+    const ev2 = ev;
+    if (dirty) {
+      ev2.returnValue = '';
+    }
+    return ev2;
   };
 
-  checkDirty = async () => {
-    if (!this.props.dirty) {
-      return;
-    }
+  useEffect(() => {
+    window.addEventListener('beforeunload', onLeave);
+    window.addEventListener('close', onLeave);
 
-    try {
-      const programToUpdate = {};
-      programToUpdate[this.props.mostRecentProgram] = {
-        code: this.props.code,
-      };
+    return () => {
+      window.removeEventListener('beforeunload', onLeave);
+      window.removeEventListener('close', onLeave);
+    };
+  }, []);
 
-      await fetch.updatePrograms(this.props.uid, programToUpdate);
-      // TODO: add functionality to be able to tell whether the fetch failed
-    } catch (err) {
-      console.log(err);
-    }
+  const openForkModal = () => {
+    setShowForkModal(true);
   };
 
-  onLeave = async (ev) => {
-    if (this.props.dirty) {
-      ev.returnValue = '';
-    }
-    return ev;
+  const closeForkModal = () => {
+    setShowForkModal(false);
   };
 
-  setCodeMirrorInstance = (codeMirrorInstance) => {
-    this.setState({ codeMirrorInstance });
-  };
+  // const checkDirty = async () => {
```

Comment:
Remove comments

Reply:
what did this used to be for? someone (probably you LOL) must've put in effort, so that was probably why @henrywangzh didn't want to remove it! Tho henry should probably remove it since you dont this it's relevant anymore :)

Reply:
I think the functionality was moved [here](https://github.com/uclaacm/TeachLAFrontend/blob/252d0b582dfe39da88b3bbb385564f17402561d0/src/components/Main.js#L68) some years ago, so this can just go